upickle/implicits/src-2/upickle/implicits/internal/Macros.scala

```diff
@@ -268,109 +300,150 @@ object Macros {
     def wrapObject(obj: Tree): Tree
 
     def wrapCaseN(companion: Tree,
-                  rawArgs: Seq[String],
-                  argSyms: Seq[Symbol],
-                  mappedArgs: Seq[String],
-                  argTypes: Seq[Type],
-                  hasDefaults: Seq[Boolean],
-                  targetType: c.Type,
-                  varargs: Boolean): Tree
+                        fields: List[Field],
+                        varargs: Boolean,
+                        targetType: c.Type): Tree
   }
 
   abstract class Reading[M[_]] extends DeriveDefaults[M] {
     val c: scala.reflect.macros.blackbox.Context
     import c.universe._
     def wrapObject(t: c.Tree) = q"new ${c.prefix}.SingletonReader($t)"
 
-    def wrapCaseN(companion: c.Tree,
-                  rawArgs: Seq[String],
-                  argSyms: Seq[Symbol],
-                  mappedArgs: Seq[String],
-                  argTypes: Seq[Type],
-                  hasDefaults: Seq[Boolean],
-                  targetType: c.Type,
-                  varargs: Boolean) = {
+    def wrapCaseN(companion: c.universe.Tree, fields: List[Field], varargs: Boolean, targetType: c.Type): c.universe.Tree = {
       val allowUnknownKeysAnnotation = targetType.typeSymbol
         .annotations
-        .find(_.tpe == typeOf[upickle.implicits.allowUnknownKeys])
-        .flatMap(_.scalaArgs.headOption)
+        .find(_.tree.tpe == typeOf[upickle.implicits.allowUnknownKeys])
+        .flatMap(_.tree.children.tail.headOption)
         .map { case Literal(Constant(b)) => b.asInstanceOf[Boolean] }
-        .headOption
 
-      val defaults = deriveDefaults(companion, hasDefaults)
+      val allFields = fields.flatMap(_.allFields).filter(_.flatten != Flatten.Map)
```

Comment:
When flattening a Map, I handled this by not assigning it an index (using -1) and treating it as an exception. And, it is not included in the field count. This approach ensures that when key-value pairs for the flattened Map are not found in the JSON, it is considered valid and does not trigger a check in checkErrorMissingKeys. (ie, it's a just empty Map)

Another approach I considered was assigning an index to the flattened Map field and using storeValueIfNotFound to create an empty Map in such cases. Which approach do you think is better?